### PR TITLE
Initial documentation

### LINKED ISSUE #1: Create project standards repo
Fresh public repo (assign this issue to that repo once it exists)
* Install Laravel project

Things to standardize (merge in comments from Motoprix week 2 walkthrough)

- Docker configuration
    - docker-compose.yml
    - changes needed in .env.example
    - everything in docker folder
- README format
- composer
    - scripts
        - scripts-descriptions?
    - tinker in dev dependencies
    - auth.json for private packages
- npm scripts
- CI configuration
- test setup
    - phpunit.xml settings and env overrides
    - base classes
    - coverage
- git ignore
- php code standards
- phpstan
- envoyer/forge setup, at least until Docker in prod is figured out
- checkly uptime monitoring

Once done, tag with a simple number like `1`, `2`, etc

### PR DESCRIPTION
**Pull request by joelclermont**
Closes #1 

The intention here is to have something easy to copy into a new project, not something to actually use as a project installer.

This doesn't match any one specific project right now. We will need to separately go through and normalize all current projects to match this format. Once we've done that, I'll tag version `1`. Consider this a baseline that could be used right now.

### CODE REVIEW

File: lang/en/auth.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?php
+
```
Comment (aaronsaray) [inline lang/en/auth.php:2]: Do we want to add these files to the phpcs path too?

File: phpunit.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
+         bootstrap="vendor/autoload.php"
+         colors="true"
+>
+    <testsuites>
+        <testsuite name="Unit">
+            <directory suffix="Test.php">./tests/Unit</directory>
+        </testsuite>
+        <testsuite name="Feature">
```
Comment (aaronsaray) [inline phpunit.xml:11]: No integration?

File: phpunit.xml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
+         bootstrap="vendor/autoload.php"
+         colors="true"
```
Comment (aaronsaray) [inline phpunit.xml:5]: Aren't we doing stop on first failure?

File: public/.htaccess [hunk 4]
```diff
@@ -0,0 +1,21 @@
+<IfModule mod_rewrite.c>
```
Comment (aaronsaray) [inline public/.htaccess:1]: Why this file? I would remove it.

File: resources/js/app.js [hunk 5]
```diff
@@ -0,0 +1 @@
+import './bootstrap';
```
Comment (aaronsaray) [inline resources/js/app.js:1]: I 100% of the time remove all of the stuff from bootstrap and put it back into this app file.  I think we should do that too.

File: routes/api.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
+
```
Comment (aaronsaray) [inline routes/api.php:2]: Should this be part of our phpcs paths?

File: routes/console.php [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+| Console Routes
+|--------------------------------------------------------------------------
+|
+| This file is where you may define all of your Closure based console
+| commands. Each Closure is bound to a command instance allowing a
+| simple approach to interacting with each command's IO methods.
+|
+*/
+
+Artisan::command('inspire', function () {
```
Comment (aaronsaray) [inline routes/console.php:17]: At least comment out. We don't want registered code if we don't need it.

File: tests/Concerns/AlertsUnwantedDBAccess.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Concerns;
+
+use Illuminate\Support\Facades\DB;
+use Tests\Exceptions\DatabaseAccessException;
+
+/**
```
Comment (aaronsaray) [inline tests/Concerns/AlertsUnwantedDBAccess.php:14]: Can remove this comment.

File: tests/Concerns/CreatingActingAs.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * Use a user model to create the acting as user
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Concerns;
+
+use App\Models\User;
+
+/**
```
Comment (aaronsaray) [inline tests/Concerns/CreatingActingAs.php:13]: Remove comment

File: tests/Concerns/CreatingActingAs.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * Trait CreatingActingAs
+ * @package Tests\Concerns
+ */
+trait CreatingActingAs
+{
+    /**
+     * @var User Current acting as user
+     */
+    protected $actingAs;
```
Comment (aaronsaray) [inline tests/Concerns/CreatingActingAs.php:22]: type hint property?

File: tests/External/.gitkeep [hunk 11]
```diff
@@ -0,0 +1 @@
+Once you add an external test, you can delete this file.
```
Comment (aaronsaray) [inline tests/External/.gitkeep:1]: https://www.aaronsaray.com/2021/better-alternative-to-gitkeep

File: tests/Feature/FeatureTestCase.php [hunk 12]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\Feature;
+
+use Tests\Concerns\CreatingActingAs;
+use Tests\Integration\IntegrationTestCase;
+
+/**
```
Comment (aaronsaray) [inline tests/Feature/FeatureTestCase.php:10]: Remove comment

**Review by aaronsaray — CHANGES_REQUESTED** [review]

**Comment by aaronsaray** [comment]
By default configuration, I couldn't get the npm to run.  This is where that default name `project` bit me because I didn't recognize it as something I had to change to be my project.

**Comment by aaronsaray** [comment]
We also talked about removing ray - should we also remove pint?

**Comment by aaronsaray** [comment]
Telescope needs to not be registered in the app config and instead locally registered only.

File: .editorconfig [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+indent_size = 4
+trim_trailing_whitespace = true
+
+[*.md]
+trim_trailing_whitespace = false
+
+[*.{yml,yaml}]
+indent_size = 2
+
+[docker-compose.yml]
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply .editorconfig:17]: Not a standard we normally tweak.

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply database/migrations/2014_10_12_000000_create_users_table.php:14]: We don't tweak migrations.

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExists('users');
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply database/migrations/2014_10_12_000000_create_users_table.php:34]: We don't tweak migrations.

File: database/migrations/2019_12_14_000001_create_personal_access_tokens_table.php [hunk 16]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply database/migrations/2019_12_14_000001_create_personal_access_tokens_table.php:7]: We don't tweak migrations.

File: docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+      - MYSQL_ROOT_PASSWORD=password
+      - MYSQL_DATABASE=app # update .env and .env.example if you change any of these values
+      - MYSQL_USER=app
+      - MYSQL_PASSWORD=app
+    ports:
+      - "${DOCKER_MYSQL_LOCAL_PORT}:3306"
+    command:
+      - "--character-set-server=utf8mb4"
+      - "--collation-server=utf8mb4_unicode_ci"
+      - "--default-authentication-plugin=mysql_native_password"
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply docker-compose.yml:37]: Check all 3 to see if they are already current defaults in mysql 8

File: routes/console.php (hunk 7 above)
Comment (joelclermont) [reply routes/console.php:17]: Not one we tweak

File: .env.example [hunk 18]
```diff
@@ -0,0 +1,63 @@
+APP_NAME=Laravel
+APP_ENV=local
+APP_KEY=
+APP_DEBUG=true
+APP_URL=https://project.test:30080
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply .env.example]: Done

File: .github/workflows/ci.yml [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@v1
+
+      - name: Setup default environment
+        run: cp .env.example .env
+
+      - name: Install mkcert
+        run: sudo apt-get update && sudo apt install libnss3-tools && brew install mkcert && mkcert -install
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply .github/workflows/ci.yml]: Done

File: .github/workflows/ci.yml [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+      - name: Generate test certificate
+        run: mkcert -cert-file docker/nginx/project.test.pem -key-file docker/nginx/project.test-key.pem project.test
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply .github/workflows/ci.yml]: Done

File: .github/workflows/ci.yml [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+      - name: Build Docker containers and start them
+        run: docker-compose -f docker-compose.yml up -d
+
+      - name: Check running containers
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply .github/workflows/ci.yml]: Removed

File: .gitignore [hunk 22]
```diff
@@ -0,0 +1,18 @@
+/node_modules
+/public/build
+/public/hot
+/public/storage
+/storage/*.key
+/vendor
+.env
+.env.backup
+.phpunit.result.cache
+Homestead.json
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply .gitignore]: Done

File: LICENSE [hunk 23]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2022 NoCompromises
+Copyright (c) 2022 No Compromises, LLC
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply LICENSE]: Removed

File: README.md [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+* `docker` folder
+  * nginx/Dockerfile - update project name for cert files
+  * nginx/nginx.conf - update project name for cert files
+* Additions and changes to `.env.example` and `.env`
+  * APP_URL - set to be https and include the Docker host port mapping
+  * COMPOSE_PROJECT_NAME
+  * DOCKER_MYSQL_LOCAL_PORT
+  * DOCKER_NGINX_LOCAL_PORT
+  * DOCKER_SERVER_NAME
+  * DB_HOST - set to use Docker mysql service
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply README.md]: I created a new Quickstart section in the docs that's easier to just follow step 1, 2, 3.

File: README.md [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+`phpunit.xml` - normal config for local and CI tests
+* generate an APP_KEY
+* set DB_DATABASE and DB_HOST for mysql-test config in docker-compose.yml
+* add a "do-not-use" for every single .env value that enables communication with an external service
+
+`phpunit-external.xml` - meant to be run separately against real external services
+* Comment out the normal external service overrides from `phpunit.xml`
+* Make sure your local .env has the correct values for running external tests
+
+Copy and overwrite the whole `tests` folder to get the right bootstrapping, test types and base classes.
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply README.md]: Done

File: docker/nginx/nginx.conf [hunk 26]
```diff
@@ -0,0 +1,19 @@
+server {
+    listen 443 ssl;
+    ssl_certificate /etc/nginx/project.test.pem;
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply docker/nginx/nginx.conf]: Done, also updated in CI

File: README.md [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+Configured with `phpstan.neon`
+
+Currently, we run all projects at PHPStan level 1. This will be increased gradually.
+
+Both applications AND tests are covered with static analysis.
+
+### Testing
+
+`phpunit.xml` - normal config for local and CI tests
+* generate an APP_KEY
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply README.md:143]: They were already set, the wording was just confusing, so I made it more explicit "this is what we did" not "this is what you should do"

File: README.md [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+
+Within the `scripts` section of `composer.json`, we make the following changes:
+* Add `ide-helper` generation to the `post-update-cmd`
+* Add `test` and `test-coverage` scripts
+* Add a dedicated `ide-helper-update` script to regenerate everything for that package
+* Add `phpcs` for detecting code standard violation
+* Add `phpcbf` for automatically fixing code standard violations - only used locally, mainly when adding a new rule
+* Add `larastan` for static analysis
+
+The `config` block has code sniffer plugins pre-allowed.
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply README.md:116]: Notes added to README, this plugin is legit.

File: composer.json [hunk 29]
```diff
@@ -0,0 +1,99 @@
+{
+    "name": "nocompromises/laravel-project-standard",
+    "type": "project",
+    "description": "The Laravel Framework.",
+    "keywords": ["framework", "laravel"],
+    "license": "MIT",
+    "require": {
+        "php": "^8.0.2",
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply composer.json]: Done

File: composer.json [hunk 30]
```diff
@@ -0,0 +1,99 @@
+{
+    "name": "nocompromises/laravel-project-standard",
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply composer.json]: Done

File: README_TEMPLATE.md [hunk 31]
```diff
@@ -0,0 +1,257 @@
+# Project/Site Name
+
+## Overview
+
+A brief description about what this project is and how it is used.
+
+Include a link to the laravel-project-standard repo and what version this project is based on.
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply README_TEMPLATE.md]: Done

File: composer.json [hunk 32]
```diff
@@ -0,0 +1,99 @@
...
+            "@php artisan ide-helper:models --write"
+        ],
+        "phpcbf": [
+            "phpcbf"
+        ],
+        "phpcs": [
+            "phpcs -s"
+        ],
+        "larastan": [
+            "@php vendor/bin/phpstan analyse"
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply composer.json]: Done

File: composer.json [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+        "laravel": {
+            "dont-discover": []
+        }
+    },
+    "config": {
+        "optimize-autoloader": true,
+        "preferred-install": "dist",
+        "sort-packages": true,
+        "allow-plugins": {
+            "dealerdirect/phpcodesniffer-composer-installer": true
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply composer.json:94]: Clarification added to README

File: config/logging.php [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+    | the box, Laravel uses the Monolog PHP logging library. This gives
+    | you a variety of powerful log handlers / formatters to utilize.
+    |
+    | Available Drivers: "single", "daily", "slack", "syslog",
+    |                    "errorlog", "monolog",
+    |                    "custom", "stack"
+    |
+    */
+
+    'channels' => [
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply config/logging.php:55]: Done

File: composer.json (hunk 32 above)
[the remark this answers is not in the document]
Comment (joelclermont) [reply composer.json:81]: Actually had to switch this back, but I added a note to the scripts section in the README

File: docker-compose.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (joelclermont) [reply docker-compose.yml]: I removed all 3 settings:
* `character-set-server` already defaults to `utf8mb4`
* `collation-server` defaults to `utf8mb4_0900_ai_ci` which is a little faster, and case/accent insensitive
* `default-authentication-plugin` is now `caching_sha2_password` by default in mysql 8, and it works without any modifications in Laravel

File: docker-compose.yml [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+    restart: unless-stopped
+
+  ## NGINX to support the app, configured to use the php-fpm-debug server
+  nginx:
+    container_name: "${COMPOSE_PROJECT_NAME}-nginx"
+    build: docker/nginx
+    working_dir: /app
+    volumes:
+      - .:/app
+      - ./storage/app/public:/app/public/storage
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply docker-compose.yml:56]: Removed

File: docker-compose.yml [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+    build:
+      context: docker/php-fpm
+      target: base
+    working_dir: /app
+    volumes:
+      - .:/app
+    depends_on:
+      - mysql
+    environment:
+      PHP_IDE_CONFIG: "serverName=${DOCKER_SERVER_NAME}"
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply docker-compose.yml:88]: Removed

File: docker/bin/test-with-coverage [hunk 37]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+source "`dirname $0`/../../.env"
+
+# when generating coverage, we need xdebug, so we must use the debug container. It's also memory hungry, so we up that limit
+docker exec -it -e "XDEBUG_MODE=coverage" "$COMPOSE_PROJECT_NAME"-php-fpm-debug php -d memory_limit=4G vendor/bin/phpunit --coverage-html=tests/html-coverage "$@"
+
+# as a convenience, we open the coverage report once it's done generating
+open "`dirname $0`/../../tests/html-coverage/index.html"
```
[the remark this answers is not in the document]
Comment (joelclermont) [reply docker/bin/test-with-coverage:9]: Removed

File: lang/en/auth.php (hunk 1 above)
Comment (joelclermont) [reply lang/en/auth.php:2]: Done

File: phpunit.xml (hunk 2 above)
Comment (joelclermont) [reply phpunit.xml:11]: Fixed

File: phpunit.xml (hunk 3 above)
Comment (joelclermont) [reply phpunit.xml:5]: Fixed

File: resources/js/app.js (hunk 5 above)
Comment (joelclermont) [reply resources/js/app.js:1]: Moved, and README updated

File: routes/api.php (hunk 6 above)
Comment (joelclermont) [reply routes/api.php:2]: Done.

File: tests/Concerns/AlertsUnwantedDBAccess.php (hunk 8 above)
Comment (joelclermont) [reply tests/Concerns/AlertsUnwantedDBAccess.php:14]: Done

File: tests/Concerns/CreatingActingAs.php (hunk 9 above)
Comment (joelclermont) [reply tests/Concerns/CreatingActingAs.php:13]: Done

File: tests/Concerns/CreatingActingAs.php (hunk 10 above)
Comment (joelclermont) [reply tests/Concerns/CreatingActingAs.php:22]: Done

File: tests/Feature/FeatureTestCase.php (hunk 12 above)
Comment (joelclermont) [reply tests/Feature/FeatureTestCase.php:10]: Done

**Comment by joelclermont** [comment]
> We also talked about removing ray - should we also remove pint?

I removed ray, pint, and sail

**Comment by joelclermont** [comment]
> Telescope needs to not be registered in the app config and instead locally registered only.

Done

File: tests/External/.gitkeep (hunk 11 above)
Comment (joelclermont) [reply tests/External/.gitkeep]: Deleted `.gitkeep` and added a README.

File: README_TEMPLATE.md [hunk 38]
```diff
@@ -85,7 +83,7 @@ You can verify that your containers are in a running state with `docker compose
 The one tool that we don't manage through Docker Compose is `node`/`npm`. The reason is that node only needs to run on demand,
 and not be constantly run in the background while we're doing development.
 
-To build the `node`/`npm` container, run: `docker build -t project-node docker/node`
+To build the `node`/`npm` container, run: `docker build -t my-project-node docker/node`
```
Comment (aaronsaray) [inline README_TEMPLATE.md:86]: Maybe I'm being dumb here but I still think there needs to be some sort of note about that you must actually build this before you try to run docker/bin/npm - it bit me and I can see myself forgetting that again.

**Review by aaronsaray — APPROVED** [review]
I would like that one line added to the README for my sake - otherwise I think this is good to go and you can merge it in at will. Great job with the updates!!

File: README_TEMPLATE.md (hunk 38 above)
Comment (joelclermont) [reply README_TEMPLATE.md]: Done

**Pull request merged**